tools/azure-devtools/src/azure_devtools/perfstress_tests/_perf_stress_runner.py

```diff
@@ -120,90 +115,140 @@ def _parse_args(self):
         self.logger.info(args)
         self.logger.info(self.per_test_args)
         self.logger.info("")
+        return test_name
 
     def _discover_tests(self, test_folder_path):
         base_classes = [PerfStressTest, BatchPerfTest, EventPerfTest]
         self._test_classes = {}
         if os.path.isdir(os.path.join(test_folder_path, 'tests')):
             test_folder_path = os.path.join(test_folder_path, 'tests')
+        sys.path.append(test_folder_path)
         self.logger.debug("Searching for tests in {}".format(test_folder_path))
 
         # Dynamically enumerate all python modules under the tests path for classes that implement PerfStressTest
-        for loader, name, _ in pkgutil.walk_packages([test_folder_path]):
+        for _, module_name, _ in pkgutil.walk_packages([test_folder_path]):
             try:
-                module = loader.find_module(name).load_module(name)
+                module = importlib.import_module(module_name)
             except Exception as e:
-                self.logger.debug("Unable to load module {}: {}".format(name, e))
+                self.logger.debug("Unable to load module {}: {}".format(module_name, e))
                 continue
             for name, value in inspect.getmembers(module):
                 if name.startswith("_"):
                     continue
                 if inspect.isclass(value):
                     if issubclass(value, _PerfTestABC) and value not in base_classes:
                         self.logger.info("Loaded test class: {}".format(name))
-                        self._test_classes[name] = value
+                        self._test_classes[name] = (value, module_name)
+
+    def _next_stage(self, title: str, track_status: bool = False, report_results: bool = False):
+        # Wait for previous stage to complete.
+        self.test_stages[title].wait()
+
+        # Stop any status tracking of the previous stage.
+        if self.status_thread.is_running:
+            self.status_thread.stop()
+
+        # If previous stage had results, report.
+        if report_results:
+            self._report_results()
+
+        self.logger.info("")
+        self.logger.info("=== {} ===".format(title))
```

Comment:
When trying this out, I noticed that if a user has log or print statements in their setup/cleanup/etc methods, these statements are sometimes not printed under the designated `=== {title} ===` section:

Example output:

```
=== Options ===
Namespace(test='DownloadTest')
Namespace(parallel=4, processes=8, duration=10, warmup=5, no_cleanup=False, sync=False, profile=False, test_proxies=None, insecure=False, max_put_size=67108864, max_block_size=4194304, buffer_threshold=4194305, client_encryption=None, max_concurrency=1, size=10240, no_client_share=False)

Global Setup for Storage Download Test...
Global Setup for Storage Download Test...
Global Setup for Storage Download Test...
Global Setup for Storage Download Test...
Setting up Storage Download Test...

=== Setup ===
Setting up Storage Download Test...
Setting up Storage Download Test...
Setting up Storage Download Test...
post-setup------
post-setup------

post-setup------
=== Post Setup ===
post-setup------

=== Warmup ===
Current         Total           Average
152             152             154.54
156             308             154.86
```

Is there an easy way to ensure that any print/log statements output under their corresponding sections?

Reply:
hmmm I'm not sure. Once line 145 has completed - each of the concurrent processes is basically allowed to proceed to its next task. Meanwhile this main thread still a few lines of code to execute before it can print the title. Even if we moved the heading logging to straight after the barrier `wait`, we couldn't guarantee that one of those processes wouldn't still get ahead of it.

Reply:
That's fine then as I don't believe many tests log/print anything in these methods anyway.The warmup and test outputs are properly placed which is more important :)